test/cpp/api/modules.cpp

```diff
@@ -550,6 +550,28 @@ TEST_F(ModulesTest, AdaptiveAvgPool3d) {
   ASSERT_EQ(y.sizes(), torch::IntArrayRef({1, 3, 3, 3}));
 }
 
+TEST_F(ModulesTest, MaxUnpool1d) {
+  auto indices = torch::tensor({{{1, 3, 4}}}, torch::kLong);
+  auto x = torch::tensor({{{2, 4, 5}}}, torch::requires_grad());
+  auto model = MaxUnpool1d{3};
+  auto y = model->forward(x, indices);
+
+  ASSERT_EQ(y.dim(), 3);
+  ASSERT_TRUE(torch::allclose(y,
+    torch::tensor({{{0, 2, 0, 4, 5, 0, 0, 0, 0}}}, torch::kFloat)));
+  ASSERT_EQ(y.sizes(), torch::IntArrayRef({1, 1, 9}));
+
+  indices = torch::tensor({{{1, 3, 4}}}, torch::kLong);
+  x = torch::tensor({{{2, 4, 5}}}, torch::requires_grad());
+  model = MaxUnpool1d{MaxUnpool1dOptions(3).stride(2).padding(1)};
+  y = model->forward(x, indices, {1, 1, 5});
+
+  ASSERT_EQ(y.dim(), 3);
+  ASSERT_TRUE(torch::allclose(y,
+    torch::tensor({{{0, 2, 0, 4, 5}}}, torch::kFloat)));
+  ASSERT_EQ(y.sizes(), torch::IntArrayRef({1, 1, 5}));
+}
```

Comment:
It would great to also test the C++ equivalent of the example in https://pytorch.org/docs/stable/nn.html#torch.nn.MaxUnpool1d, to show that MaxPool -> MaxUnpool works well.

Reply:
Do you think we can add the test for MaxPool -> MaxUnpool? Maybe something like:
```cpp
  {
    MaxPool1d pool(MaxPool1dOptions(2).stride(2));
    MaxUnpool1d unpool(MaxUnpool1dOptions(2).stride(2));
    auto input = torch::tensor({{{1., 2., 3., 4., 5., 6., 7., 8.}}});
    torch::Tensor output, indices;
    std::tie(output, indices) = pool->forward_with_indices(input);
    ASSERT_TRUE(torch::allclose(
      unpool(output, indices),
      torch::tensor({{{0., 2., 0., 4., 0., 6., 0., 8.}}})));
  }
  {
    MaxPool1d pool(MaxPool1dOptions(2).stride(2));
    MaxUnpool1d unpool(MaxUnpool1dOptions(2).stride(2));
    auto input = torch::tensor({{{1., 2., 3., 4., 5., 6., 7., 8., 9.}}});
    torch::Tensor output, indices;
    std::tie(output, indices) = pool->forward_with_indices(input);
    ASSERT_TRUE(torch::allclose(
      unpool(output, indices, /*output_size=*/input.sizes()),
      torch::tensor({{{0., 2., 0., 4., 0., 6., 0., 8., 0.}}})));
    ASSERT_TRUE(torch::allclose(
      unpool(output, indices),
      torch::tensor({{{0., 2., 0., 4., 0., 6., 0., 8.}}})));
  }
```